// fast-readdir/index.js
const fs = require('fs')
const join = require('path').join
const util = require('util')
const readdir = util.promisify(fs.readdir)
const stat = util.promisify(fs.stat)

exports.freaddir = async function freaddir(path) {
  let out = {}

  let files
  try {
    files = await readdir(path)
  } catch(ex) {
    return out
  }

  for (let fileName of files) {
    let filePath = join(path, fileName)
    let isDir
    try {
      isDir = (await stat(filePath)).isDirectory()
    } catch(ex) {
      continue
    }

    if (isDir) {
      let children = await freaddir(filePath)
      out[filePath] = {
        name: fileName,
        path: filePath,
        isDir: true,
        children: children,
      }
    } else {
      out[filePath] = {
        name: fileName,
        path: filePath,
        isDir: false,
      }
    }
  }

  return out
}

exports.freaddirSync = function freaddirSync(path) {
  let out = {}

  let files
  try {
    files = fs.readdirSync(path)
  } catch(ex) {
    return out
  }

  for (let fileName of files) {
    let filePath = join(path, fileName)
    let isDir
    try {
      isDir = fs.statSync(filePath).isDirectory()
    } catch(ex) {
      continue
    }

    if (isDir) {
      let children = freaddirSync(filePath)
      out[filePath] = {
        name: fileName,
        path: filePath,
        isDir: true,
        children: children,
      }
    } else {
      out[filePath] = {
        name: fileName,
        path: filePath,
        isDir: false,
      }
    }
  }

  return out
}
